### PR TITLE
3rdparty SPIRV/GLSL upgrade and SkinningBenchmark subgroup bone fetch

### PR DESCRIPTION
**Pull request by Crisspl**

### CODE REVIEW

**Comment by devshgraphicsprogramming** [comment]
can I merge or do I still need to wait on something?

**Comment by devshgraphicsprogramming** [comment]
btw you havent changed the 2 submodule repos to my forks.

**Comment by Crisspl** [comment]
Opened against nabla as well

**Comment by devshgraphicsprogramming** [comment]
I'll merge it here, tonight

**Comment by Crisspl** [comment]
Our contribution got merged into spirv-cross. But we're still using our spirv-cross repo as a submodule, right? Then what should I do? Pull and set it on primary repo's `HEAD`? Obviously `2020-09-17` tag you skaded for before doesnt support subgroups.

**Comment by Crisspl** [comment]
Ran the benchmark: (RTX 2070)
```
Benchmark for case nr. 0
Result 0: 1945.325568ms.
Result 1: 1961.448320ms.
Result 2: 1970.530240ms.
Result 3: 1972.837760ms.
Result 4: 1975.484416ms.
Avg time: 1965.125366

Benchmark for case nr. 1
Result 0: 1970.074048ms.
Result 1: 1973.327872ms.
Result 2: 1980.258304ms.
Result 3: 1981.993216ms.
Result 4: 1984.545184ms.
Avg time: 1978.039673

Benchmark for case nr. 2
Result 0: 2006.147072ms.
Result 1: 2009.796608ms.
Result 2: 2010.246208ms.
Result 3: 2011.929376ms.
Result 4: 2012.337056ms.
Avg time: 2010.091187

Benchmark for case nr. 3
Result 0: 2171.191680ms.
Result 1: 2176.405504ms.
Result 2: 2176.366464ms.
Result 3: 2176.658496ms.
Result 4: 2177.943008ms.
Avg time: 2175.713135

Benchmark for case nr. 4
Result 0: 1998.852192ms.
Result 1: 2001.749440ms.
Result 2: 2003.395200ms.
Result 3: 2003.919648ms.
Result 4: 2004.760576ms.
Avg time: 2002.535400
```

**Comment by devshgraphicsprogramming** [comment]
> Our contribution got merged into spirv-cross. But we're still using our spirv-cross repo as a submodule, right? Then what should I do? Pull and set it on primary repo's `HEAD`? Obviously `2020-09-17` tag you skaded for before doesnt support subgroups.

we will always use our own fork, we'll add more features to it.

File: examples_tests/31.SkinningDataBenchmark/5.vert [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    GET_NORMAL_MATRIX_COL(0);
+    GET_NORMAL_MATRIX_COL(1);
+    GET_NORMAL_MATRIX_COL(2);
+    GET_NORMAL_MATRIX_COL(3);
+
+    return retval;
+}
+#define COALESCING_DWORDS_LOG2 4 // GCN can fetch only 64bytes in a single request
+#define SUBGROUP_THRESH 16
```
Comment (devshgraphicsprogramming) [inline examples_tests/31.SkinningDataBenchmark/5.vert:56]: change the theshold to `(0x1<<COALESCING_DWORDS_LOG2)`

Also does changing the `COALESCING_DWORDS_LOG2` to 5 and deleting the `notEnoughInvocations` (always false) change the perf of the benchmark much?

File: examples_tests/31.SkinningDataBenchmark/5.vert [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+		  for (int j=0; j<SUBGROUP_THRESH; j++)
+			retval.data[oit++] = subgroupBroadcast(tmp,iit++);
+		}
+    }
+
+    return toBoneData(retval);
+  }
+  else
+//#endif
+  return boneSSBO_structs.data[_boneID];
```
Comment (devshgraphicsprogramming) [inline examples_tests/31.SkinningDataBenchmark/5.vert:108]: wonder if we hoisted the `BoneData_dword retval` above the if, and had only one SSBO with the `restrict` keyword, how the perf would change then (both branches do `toBoneData`

**Comment by devshgraphicsprogramming** [comment]
Nvidia 1050 Mobile
```
Result 0: 3188.719616ms.
Result 1: 3277.758464ms.
Result 2: 2831.123456ms.
Result 3: 3004.645344ms.
Result 4: 2987.333632ms.
Avg time: 3057.916260

Benchmark for case nr. 1
Result 0: 3437.733888ms.
Result 1: 3374.459904ms.
Result 2: 2657.101824ms.
Result 3: 3049.407488ms.
Result 4: 3201.132544ms.
Avg time: 3143.967285

Benchmark for case nr. 2
Result 0: 2906.318848ms.
Result 1: 2935.910400ms.
Result 2: 2532.075520ms.
Result 3: 2534.433792ms.
Result 4: 2530.915328ms.
Avg time: 2687.930664

Benchmark for case nr. 3
Result 0: 2635.242496ms.
Result 1: 2642.153472ms.
Result 2: 2617.835520ms.
Result 3: 2631.688192ms.
Result 4: 2640.667648ms.
Avg time: 2633.517578

Benchmark for case nr. 4
Result 0: 2808.863744ms.
Result 1: 2595.141632ms.
Result 2: 2583.285760ms.
Result 3: 2639.542208ms.
Result 4: 2582.702080ms.
Avg time: 2641.906982
```

**Comment by Crisspl** [comment]
`COALESCING_DWORDS_LOG2=5`, RTX 2070 (case 4)
```
Benchmark for case nr. 0
Result 0: 1945.659840ms.
Result 1: 1971.484864ms.
Result 2: 1981.332544ms.
Result 3: 2020.957504ms.
Result 4: 2009.023296ms.
Avg time: 1985.691650

Benchmark for case nr. 1
Result 0: 2028.918784ms.
Result 1: 2001.373184ms.
Result 2: 2107.480064ms.
Result 3: 2060.787488ms.
Result 4: 2058.666272ms.
Avg time: 2051.445068

Benchmark for case nr. 2
Result 0: 2069.369216ms.
Result 1: 2071.384064ms.
Result 2: 2073.463840ms.
Result 3: 2075.557632ms.
Result 4: 2077.211008ms.
Avg time: 2073.397217

Benchmark for case nr. 3
Result 0: 2236.676704ms.
Result 1: 2235.980352ms.
Result 2: 2238.579936ms.
Result 3: 2256.968288ms.
Result 4: 2259.242688ms.
Avg time: 2245.489990

Benchmark for case nr. 4
Result 0: 2073.350144ms.
Result 1: 2062.793280ms.
Result 2: 2061.597216ms.
Result 3: 2066.504992ms.
Result 4: 2067.678240ms.
Avg time: 2066.385010
```

**Comment by Crisspl** [comment]
`COALESING_DWORDS_LOG2=5` and just one `restrict` buffer, RTX 2070 (case 4)
```
Benchmark for case nr. 0
Result 0: 1971.348160ms.
Result 1: 1986.811968ms.
Result 2: 2000.763232ms.
Result 3: 2056.366336ms.
Result 4: 2094.850048ms.
Avg time: 2022.027954

Benchmark for case nr. 1
Result 0: 2171.216416ms.
Result 1: 2178.971520ms.
Result 2: 2187.540608ms.
Result 3: 2182.462400ms.
Result 4: 2164.656160ms.
Avg time: 2176.969482

Benchmark for case nr. 2
Result 0: 2094.723776ms.
Result 1: 2066.329440ms.
Result 2: 2066.724864ms.
Result 3: 2065.878944ms.
Result 4: 2064.435200ms.
Avg time: 2071.618652

Benchmark for case nr. 3
Result 0: 2224.043520ms.
Result 1: 2228.813824ms.
Result 2: 2240.063488ms.
Result 3: 2227.519488ms.
Result 4: 2214.503968ms.
Avg time: 2226.989014

Benchmark for case nr. 4
Result 0: 2053.507392ms.
Result 1: 2063.912960ms.
Result 2: 2123.744768ms.
Result 3: 2061.885440ms.
Result 4: 2050.065984ms.
Avg time: 2070.623535
```

**Comment by Hazardu** [comment]
GTX 1060 6GB
```
Benchmark for case nr. 0
Result 0: 4046.638080ms.
Result 1: 4047.713280ms.
Result 2: 4059.796480ms.
Result 3: 4154.593216ms.
Result 4: 4055.063552ms.
Avg time: 4072.760986

Benchmark for case nr. 1
Result 0: 4052.248576ms.
Result 1: 4050.157312ms.
Result 2: 4048.704512ms.
Result 3: 4051.081216ms.
Result 4: 4046.565376ms.
Avg time: 4049.751221

Benchmark for case nr. 2
Result 0: 4043.850752ms.
Result 1: 4043.276288ms.
Result 2: 4043.407360ms.
Result 3: 4069.681152ms.
Result 4: 4070.952960ms.
Avg time: 4054.233643

Benchmark for case nr. 3
Result 0: 4070.839296ms.
Result 1: 4090.315776ms.
Result 2: 4104.312832ms.
Result 3: 4100.513792ms.
Result 4: 4098.144256ms.
Avg time: 4092.824951

Benchmark for case nr. 4
Result 0: 4101.523456ms.
Result 1: 4102.410208ms.
Result 2: 4106.982400ms.
Result 3: 4101.438464ms.
Result 4: 4100.418560ms.
Avg time: 4102.554199
```

**Comment by devshgraphicsprogramming** [comment]
ok seems that subgroup tricks dont win us extra perf.

**Pull request merged**